### PR TITLE
Improve discovery logic

### PR DESCRIPTION
**Pull request by asottile**
Should be ready to review.

New awesomeness:
- Errors go to stderr (WHOA)
- Real stacktraces on errors (WHOA)
- Discovery by directory works `testify yelp/tests/selenium`
- Code is shorter!

### CODE REVIEW

File: test/discovery_failure_test.py [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
-class BrokenImportTestCase(TestCase):
+HERE = os.path.dirname(os.path.abspath(__file__))
+
+
+def cmd_output(*cmd):
+    proc = subprocess.Popen(
+        cmd, cwd=HERE, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+    )
+    stdout, stderr = proc.communicate()
+    assert proc.returncode != 0
```
Comment (bukzor) [inline test/discovery_failure_test.py]: `assert ..., (stdout, stderr)` plz, so we have something to debug if it fails.

Imo assert should always have a second arg.

File: test/discovery_failure_test.py [hunk 2]
```diff
@@ -38,12 +49,10 @@ def delete_broken_import_file(self):
...
                 _log.error("Could not remove broken import file %s: %r" % (f, exc))
 
-    @class_setup
+    @T.class_setup_teardown
     def setup_import_file(self):
         self.create_broken_import_file()
-
-    @class_teardown
-    def teardown_import_file(self):
+        yield
```
Comment (bukzor) [inline test/discovery_failure_test.py]: try/finally?

File: test/discovery_failure_test.py [hunk 3]
```diff
@@ -51,32 +60,30 @@ class DiscoveryFailureTestCase(BrokenImportTestCase):
...
+        T.assert_in('DISCOVERY FAILURE', stdout)
+        T.assert_in('ImportError: No module named non_existent_module', stderr)
 
 
 class DiscoveryFailureUnknownErrorTestCase(BrokenImportTestCase):
-    @class_setup
+    @T.class_setup_teardown
     def setup_import_file(self):
         self.create_broken_import_file(contents='raise AttributeError("aaaaa!")')
+        yield
```
Comment (bukzor) [inline test/discovery_failure_test.py]: (same)?

File: test/test_case_test.py [hunk 4]
```diff
@@ -149,11 +149,15 @@ class UnitTestTestYoDawg(TestCase):
     """Make sure we actually detect and run all steps in unittest.TestCases."""
 
     def test_unit_test_status(self):
-        runner = test_runner.TestRunner(UnitTest)
+        runner = test_runner.TestRunner(
+            UnitTest.__module__ + ' ' + UnitTest.__name__,
```
Comment (bukzor) [inline test/test_case_test.py]: String concat is a smell.
Usually we'd want to pass the data unmunged.

File: test/test_discovery_test.py [hunk 5]
```diff
@@ -13,57 +7,35 @@
...
         # Exhaust the generator to catch exceptions
         return [mod for mod in test_discovery.discover(path)]
 
 
-def relative(func):
-    'decorator for tests that rely on relative paths'
-    @wraps(func)
-    def wrapped(*args, **kwargs):
-        cwd = getcwd()
-        chdir(HERE)
```
Comment (bukzor) [inline test/test_discovery_test.py]: still need a fixture with chdir() no?

File: test/discovery_failure_test.py (hunk 2 above)
Comment (asottile) [inline test/discovery_failure_test.py]: setup_teardown is equivalent to try finally without needing to write try finally

File: test/test_discovery_test.py (hunk 5 above)
Comment (asottile) [inline test/test_discovery_test.py]: nope :D

File: testify/test_discovery.py [hunk 6]
```diff
@@ -31,18 +30,48 @@ class DiscoveryError(TestifyError):
...
-def gather_test_paths(testing_dir):
-    """Given a directory path, yield up paths for all py files inside of it"""
-    for adir, subdirs, subfiles in os.walk(testing_dir):
-        # ignore .svn directories and other such hiddens
-        if adir.startswith('.'):
+def to_module(path):
+    path = os.path.relpath(path)
+
+    if path.startswith('..'):
+        raise DiscoveryError('Test outside of current directory')
```
Comment (bukzor) [inline testify/test_discovery.py]: This is starting to converge with some pytest code which I helped fix recently. Maybe you will find it useful:

https://bitbucket.org/logilab/pylint/src/c4231fff3e3429f76e07e9762d51f312c5322ede/lint.py?at=default#cl-1054

https://bitbucket.org/logilab/pylint/src/c4231fff3e3429f76e07e9762d51f312c5322ede/lint.py?at=default#cl-81

Here's the PR. Maybe it's more readable: https://bitbucket.org/logilab/pylint/commits/58c66aa083777059a2e6b46f6a0545a2f4977097

Comment (bukzor) [inline testify/test_discovery.py]: I don't see any particular reason for this case to be an error, but as long as it's not a regression relative to the parent changeset, it doesn't really matter.

File: test/discovery_failure_test.py [hunk 7]
```diff
@@ -1,82 +1,77 @@
...
-            discovered_tests = test_discovery.discover(self.broken_import_module)
-            discovered_tests.next()
-        except DiscoveryError as exc:
-            assert_in('Got unknown error when trying to import', str(exc))
-        else:
-            assert False, 'Expected DiscoveryError.'
+        stdout, stderr = cmd_output(
+            'python', '-m', 'testify.test_program', self.broken_import_module,
+        )
+        T.assert_in('DISCOVERY FAILURE', stdout)
```
Comment (bukzor) [inline test/discovery_failure_test.py]: I'd really rather see the whole string assert_equal'd, unless there's too many bits that vary.

**Pull request merged**